rbd: add 'mirror snapshot schedule' commands

trociny opened the pull request:
`rbd` cli for mirror snapshot scheduler implemented in #32434

https://tracker.ceph.com/issues/41559

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

trociny commented:
@dillaman It is for discussion.

In this implementation a schedule level is specified via pool, namespace and image command line options (no options means the global level).

When listing schedules I think the user may have two goals: 1)  to see what schedules are configured on some specific level (e.g. on 'rbd' pool level), and 2) to see how images are going to be scheduled e.g. for 'rbd' pool, or 'rbd/ns1' namespace, which depends not only how this level is configured but also how parent and child levels are configured too. For the first case the `list` command is used, for the second case the `list --recursive` is used.

It looks clear for me, but I am not sure how clear it would look for users. Below is an example to illustrate:
```
% rbd mirror snapshot schedule add 1h 00:20 # add a global level schedule
% rbd mirror snapshot schedule add 3h # add one more global level schedule
% rbd mirror snapshot schedule add -p rbd 2h # add rbd pool level schedule
% rbd mirror snapshot schedule add --image rbd/test 30m 00:10 # add rbd/test image level schedule
% rbd mirror snapshot schedule add -p rbd --namespace ns1 4h # add rbd/ns1 namespace schedule
% rbd mirror snapshot schedule status # show scheduled images
NEXT SCHEDULED TIME IMAGE
2020-01-26 14:40:00 rbd/test
2020-01-26 16:00:00 rbd/image
2020-01-26 16:00:00 rbd/ns1/test1
% rbd mirror snapshot schedule ls # list configured schedules on the global level
every 3h, every 1h starting at 00:20:00
% rbd mirror snapshot schedule ls -p rbd # list configured schedules on the rbd pool level
every 2h
% rbd mirror snapshot schedule ls -R # dump (list recurcively) all schedules
POOL NAMESPACE IMAGE SCHEDULE                                
-    -         -     every 3h, every 1h starting at 00:20:00 
rbd  -         -     every 2h                                
rbd  ns1       -     every 4h                                
rbd            test  every 30m starting at 00:10:00          
% rbd mirror snapshot schedule ls -p rbd -R # list all schedules that apply to images from rbd pool
POOL NAMESPACE IMAGE SCHEDULE                       
rbd  -         -     every 2h                       
rbd  ns1       -     every 4h                       
rbd            test  every 30m starting at 00:10:00 
% rbd mirror snapshot schedule ls --image rbd/ns1/test1 -R # list schedules that apply to rbd/ns1/test1 image
POOL NAMESPACE IMAGE SCHEDULE 
rbd  ns1       -     every 4h 
```
Also, right now for `ceph rbd mirror snapshot schedule status` command the rbd_support module returns plain output and the rbd wrapper just adds a header to it. But now I am thinking, if in the future we want to extend somehow the status output, it might create some compatibility issues when using an older rbd client. So, may be change the rbd_support module to output in jason format? It would mean some additional work now to parse it on `rbd` side, but it would be much easier to extend later (an older rbd would just ignore unknown fields). Or the header could be added by the rbd_suspport module and and the `rbd` wrapper would just print everything it gets from the mgr.

dillaman commented:
> When listing schedules I think the user may have two goals: 1) to see what schedules are configured on some specific level (e.g. on 'rbd' pool level), and 2) to see how images are going to be scheduled e.g. for 'rbd' pool, or 'rbd/ns1' namespace, which depends not only how this level is configured but also how parent and child levels are configured too. For the first case the `list` command is used, for the second case the `list --recursive` is used.

Does it actually inherit from lower levels or does it just take the nearest level as truth? For a simplicity point of view, if someone lists for pool, namespace, or image, shouldn't it always just show the "recursive" schedule so long as it at least tells the user which level the schedule is coming from?

> ```
> % rbd mirror snapshot schedule add 1h 00:20 # add a global level schedule
> % rbd mirror snapshot schedule add 3h # add one more global level schedule
> % rbd mirror snapshot schedule add -p rbd 2h # add rbd pool level schedule
> % rbd mirror snapshot schedule add --image rbd/test 30m 00:10 # add rbd/test image level schedule
> % rbd mirror snapshot schedule add -p rbd --namespace ns1 4h # add rbd/ns1 namespace schedule
> % rbd mirror snapshot schedule status # show scheduled images
> NEXT SCHEDULED TIME IMAGE
> 2020-01-26 14:40:00 rbd/test
> 2020-01-26 16:00:00 rbd/image
> 2020-01-26 16:00:00 rbd/ns1/test1
> % rbd mirror snapshot schedule ls # list configured schedules on the global level
> every 3h, every 1h starting at 00:20:00
> % rbd mirror snapshot schedule ls -p rbd # list configured schedules on the rbd pool level
> every 2h
> % rbd mirror snapshot schedule ls -R # dump (list recurcively) all schedules
> POOL NAMESPACE IMAGE SCHEDULE                                
> -    -         -     every 3h, every 1h starting at 00:20:00 
> rbd  -         -     every 2h                                
> rbd  ns1       -     every 4h                                
> rbd            test  every 30m starting at 00:10:00          
> % rbd mirror snapshot schedule ls -p rbd -R # list all schedules that apply to images from rbd pool
> POOL NAMESPACE IMAGE SCHEDULE                       
> rbd  -         -     every 2h                       
> rbd  ns1       -     every 4h                       
> rbd            test  every 30m starting at 00:10:00 
> % rbd mirror snapshot schedule ls --image rbd/ns1/test1 -R # list schedules that apply to rbd/ns1/test1 image
> POOL NAMESPACE IMAGE SCHEDULE 
> rbd  ns1       -     every 4h 
> ```

In the example above, does "rbd/test" get a mirror snapshot every 30m (from image) + every 2hrs (from pool)? Either way, it seems like the `-R` optional might not be worthwhile since you should see the schedule(s) that apply to whatever level you specified, right? 

> Also, right now for `ceph rbd mirror snapshot schedule status` command the rbd_support module returns plain output and the rbd wrapper just adds a header to it. But now I am thinking, if in the future we want to extend somehow the status output, it might create some compatibility issues when using an older rbd client. So, may be change the rbd_support module to output in jason format? 

Sounds good to me

trociny commented:
> Does it actually inherit from lower levels or does it just take the nearest level as truth?

I takes the nearest level as truth (if I correctly understand what you mean). If a schedule is set on a level, it overrides all schedules on the levels above.

> In the example above, does "rbd/test" get a mirror snapshot every 30m (from image) + every 2hrs (from pool)?

Only "every 30m".

So, if I understand you correctly, you want we remove '-R' option and for `ls` we have something like this (taking the configuration from the example above):
```
% rbd mirror snapshot schedule ls # global level configuration
POOL NAMESPACE IMAGE SCHEDULE                                
-    -         -     every 3h, every 1h starting at 00:20:00 

% rbd mirror snapshot schedule ls -p test # 'test' pool level configuration, global level is show, because the 'test' pool level does not have an override.
POOL NAMESPACE IMAGE SCHEDULE                                
-    -         -     every 3h, every 1h starting at 00:20:00 

% rbd mirror snapshot schedule ls --image rbd/test # 'rbd/test' image level configuration
POOL NAMESPACE IMAGE SCHEDULE                                
rbd            test  every 30m starting at 00:10:00          

% rbd mirror snapshot schedule ls --image rbd/ns1/test1 # 'rbd/ns1/test1' image level configuration. the 'rbd/ns1' namespace level is shown, because the image level does not have an override
POOL NAMESPACE IMAGE SCHEDULE 
rbd  ns1       -     every 4h 
```

I am fine with this (for non recursive case). But may be still leave the recursive option, to output it as it is now? Because a user might want to see the "whole picture", e.g. how all images in the pool are going to be scheduled. And for this, without recursive option, she would need to probe every possible levels (pool, all namespaces, all images in all namespaces). I understand it is somewhat similar to what we have for pool/image level config overrides, and we don't have recursive option there. But this case seems a little different, and have a possibility to see the whole schedule table seems important.

Anyway, if you think the `-R` is not necessary, then for not recursive `ls` output, when we always have only one row in the table, probably we need to change it from the table oriented to the record oriented, something like this:
```
% rbd mirror snapshot schedule ls # global level configuration
pool: -
namespace: -
image: -
schedule: every 3h, every 1h starting at 00:20:00
```

dillaman commented:
> I am fine with this (for non recursive case). But may be still leave the recursive option, to output it as it is now? Because a user might want to see the "whole picture", e.g. how all images in the pool are going to be scheduled. And for this, without recursive option, she would need to probe every possible levels (pool, all namespaces, all images in all namespaces). I understand it is somewhat similar to what we have for pool/image level config overrides, and we don't have recursive option there. But this case seems a little different, and have a possibility to see the whole schedule table seems important.

Maybe annotate the parent level rows to indicate that those scheduled won't be applied / they are overridden since otherwise it could be confusing.

> Anyway, if you think the `-R` is not necessary, then for not recursive `ls` output, when we always have only one row in the table, probably we need to change it from the table oriented to the record oriented, something like this:

You could have more than one schedule applied at a given level, right? So the table still makes sense to me.

trociny commented:
> Maybe annotate the parent level rows to indicate that those schedules won't be applied / they are overridden since otherwise it could be confusing.

I am not quite sure I understand. In my current implementation `ls -R` outputs only schedules that will be applied. See my examples for `ls -p rbd -R` and `ls --image rbd/ns1/test1 -R`: if there is no a schedule on this level, then the first parent (which will apply) is displayed.

> You could have more than one schedule applied at a given level, right? So the table still makes sense to me.

If there is more than one schedule they still display in one column, comma separated (see the example for global level configuration).

dillaman commented:
> > Maybe annotate the parent level rows to indicate that those schedules won't be applied / they are overridden since otherwise it could be confusing.
> 
> I am not quite sure I understand. In my current implementation `ls -R` outputs only schedules that will be applied. See my examples for `ls -p rbd -R` and `ls --image rbd/ns1/test1 -R`: if there is no a schedule on this level, then the first parent (which will apply) is displayed.

So in that case, why not just drop `-R` and always show what schedules will apply? The output already includes from which level the schedule comes from.

> > You could have more than one schedule applied at a given level, right? So the table still makes sense to me.
> 
> If there is more than one schedule they still display in one column, comma separated (see the example for global level configuration).

I wouldn't change the output format based upon if you have 1 or >1 schedule.

trociny commented:
> So in that case, why not just drop -R and always show what schedules will apply? The output already includes from which level the schedule comes from.

Citing myself:

> When listing schedules I think the user may have two goals: 1) to see what schedules are configured on some specific level (e.g. on 'rbd' pool level), and 2) to see how images are going to be scheduled e.g. for 'rbd' pool, or 'rbd/ns1' namespace, 

For (1) the "recursive" output would not be necessary, but sure one can see it from the "recursive" output too. So, if you don't like `-R` option I am fine with your suggestion. 

Just to make sure, `rbd mirror snapshot schedule ls` command will output schedules recurcively for all levels starting from the current and below? I.e. for the configuration we had in the example above, the output would be:
```
% rbd mirror snapshot schedule ls # list all schedules
POOL NAMESPACE IMAGE SCHEDULE                                
-    -         -     every 3h, every 1h starting at 00:20:00 
rbd  -         -     every 2h                                
rbd  ns1       -     every 4h                                
rbd            test  every 30m starting at 00:10:00          
% rbd mirror snapshot schedule ls -p rbd # list all schedules that apply to images from rbd pool
POOL NAMESPACE IMAGE SCHEDULE                       
rbd  -         -     every 2h                       
rbd  ns1       -     every 4h                       
rbd            test  every 30m starting at 00:10:00 
% rbd mirror snapshot schedule ls --image rbd/ns1/test1 # list schedules that apply to rbd/ns1/test1 image
POOL NAMESPACE IMAGE SCHEDULE 
rbd  ns1       -     every 4h 
```
Also, please note how two schedules on the global level are displayed:
```
POOL NAMESPACE IMAGE SCHEDULE                                
-    -         -     every 3h, every 1h starting at 00:20:00 
```
Or do you want we have two rows for this?
```
POOL NAMESPACE IMAGE SCHEDULE                                
-    -         -     every 3h
-    -         -     every 1h starting at 00:20:00 
```
This seems to be confusing to me.

BTW, it seems `ceph rbd mirror snapshot schedule list` command we added to the mgr is not very useful, as `rbd` is going to use only `dump`. Then maybe remove `ceph rbd mirror schedule list` and rename `dump` to list?

dillaman commented:
> Just to make sure, `rbd mirror snapshot schedule ls` command will output schedules recurcively for all levels starting from the current and below? I.e. for the configuration we had in the example above, the output would be:
> 
> ```
> % rbd mirror snapshot schedule ls # list all schedules
> POOL NAMESPACE IMAGE SCHEDULE                                
> -    -         -     every 3h, every 1h starting at 00:20:00 
> rbd  -         -     every 2h                                
> rbd  ns1       -     every 4h                                
> rbd            test  every 30m starting at 00:10:00          
> % rbd mirror snapshot schedule ls -p rbd # list all schedules that apply to images from rbd pool
> POOL NAMESPACE IMAGE SCHEDULE                       
> rbd  -         -     every 2h                       
> rbd  ns1       -     every 4h                       
> rbd            test  every 30m starting at 00:10:00 
> % rbd mirror snapshot schedule ls --image rbd/ns1/test1 # list schedules that apply to rbd/ns1/test1 image
> POOL NAMESPACE IMAGE SCHEDULE 
> rbd  ns1       -     every 4h 
> ```

Perhaps "rbd mirror snapshot schedule ls" vs "rbd mirror snapshot schedule get" to differentiate between the "dump"-like recursive behavior vs get the schedules for a specific level? 

> Also, please note how two schedules on the global level are displayed:
> ...
> Or do you want we have two rows for this?

The one nice thing about two rows is that it helps you clearly understand how to delete a specific schedule entry. 

> BTW, it seems `ceph rbd mirror snapshot schedule list` command we added to the mgr is not very useful, as `rbd` is going to use only `dump`. Then maybe remove `ceph rbd mirror schedule list` and rename `dump` to `list`?

Not that we would be exposing users to our MGR commands, but would it be consistent w/ CephFS's needs?

trociny commented:
> Perhaps "rbd mirror snapshot schedule ls" vs "rbd mirror snapshot schedule get" to differentiate between the "dump"-like recursive behavior vs get the schedules for a specific level?

We discussed `get` vs `ls` in #32434. My rationale to use `ls` was that it may return several schedules, and in this case `ls` looked more natural.  @jan--f seemed to agree with me. Actually I don't have strong objection against `get`. On the other hand I don't see why having `get` + `ls` is better than `ls` and `ls -R`. The `ls` + `ls -R` looks natural to me because it mimics the behavior of the posix fs `ls` utility. Anyway, any solution is good to me, we just need to coordinate it with cephfs.

> The one nice thing about two rows is that it helps you clearly understand how to delete a specific schedule entry.

Agree, but it seems to make more difficult to analyze "recursive" (dump like) output. Here is an example of "all level schedules in one row" format:
```
POOL NAMESPACE IMAGE SCHEDULE                                
-    -         -     every 3h, every 1h starting at 00:20:00 
rbd  -         -     every 2h                                
rbd  ns1       -     every 4h, every 1 d starting at 02:00:00
rbd            test  every 30m starting at 00:10:00          
```
And how it would look in "multiple rows" format:
```
POOL NAMESPACE IMAGE SCHEDULE                                
-    -         -     every 3h
-    -         -     every 1h starting at 00:20:00 
rbd  -         -     every 2h                                
rbd  ns1       -     every 4h
rbd  ns1       -     every 1 d starting at 02:00:00
rbd            test  every 30m starting at 00:10:00          
```
In the first case you know that every row below overrides the schedules above. In the second case it is not true. Also you may not notice that a level has two rows and think it is has only one schedule. But it is subjective.

> Not that we would be exposing users to our MGR commands, but would it be consistent w/ CephFS's needs?

I would not care much about the MGR commands, as the users are not supposed to use them. I think these are the `rbd` commans we should be care abut, to make sure it looks consistent as much as possible (both for command names and their output) with what cephfs has.

May be @jan--f will have some comments about this? Do you have a command to list all schedules recursively? How does `ls` output looks in your case?

dillaman commented:
> The `ls` + `ls -R` looks natural to me because it mimics the behavior of the posix fs `ls` utility. Anyway, any solution is good to me, we just need to coordinate it with cephfs.

Fair enough.

> Agree, but it seems to make more difficult to analyze "recursive" (dump like) output. Here is an example of "all level schedules in one row" format:

Fair enough. Given that the XML/json output break it up, that satisfies my desire to have a good mapping back to the actual add/rm calls that resulted in the schedule.

> May be @jan--f will have some comments about this? Do you have a command to list all schedules recursively? How does `ls` output looks in your case?

Pending this feedback, if we can drop the extra command, that would be great.

dillaman commented on src/librbd/api/Mirror.h:
```diff
@@ -82,6 +90,15 @@ struct Mirror {
   static void image_get_info(ImageCtxT *ictx,
                              mirror_image_info_t *mirror_image_info,
                              Context *on_finish);
+  static int image_get_info(librados::IoCtx& io_ctx,
```
Nit: since this doesn't seem to be used, can we drop it for now?

trociny replied:
Again, it is already merged to the master. And it is used internally in Mirror.cc. Certainly we could live without it, just using the async version in Mirror.cc, though personally I like as it is now. But I can push a PR (or add a commit to this PR) to drop it, if you still think it is a good idea.

dillaman replied:
If it's used, it's fine.

dillaman commented on src/test/cli/rbd/help.t:
```diff
@@ -5,138 +5,155 @@
+      mirror pool status                   Show status for all mirrored images in
+                                           the pool.
+      mirror snapshot schedule add         Add mirror snapshot schedule.
+      mirror snapshot schedule list (... ls)
```
Nit: might be nice to adjust the wrap point down so that we don't start putting most of the config actions on two lines above.

dillaman commented on src/include/rbd/librbd.h:
```diff
@@ -587,6 +587,14 @@ CEPH_RBD_API int rbd_mirror_image_instance_id_list(rados_ioctx_t io_ctx,
                                                             size_t len);
+CEPH_RBD_API int rbd_mirror_image_info_list(
+    rados_ioctx_t io_ctx, rbd_mirror_image_mode_t *mode_filter,
+    const char *start_id, size_t max, char **image_ids,
```
Nit: would it make sense to consolidate `max` and `num_entries` to a single in/out parameter?

trociny replied:
@dillaman This change was from #32434 this PR was rebased on, because at that moment #32434 was not still merged. Now as #32434 is merged this change is in the master already.

I suppose we could  consolidate `max` and `num_entries` as an additional PR, though note, that all other `rbd_mirror_image_*_list` functions have both params, so it would be inconsistent.

dillaman replied:
No worries, I forgot it was part of that other PR.

dillaman commented on src/tools/rbd/action/MirrorSnapshotSchedule.cc:
```diff
@@ -0,0 +1,607 @@
+int parse_schedule_name(const std::string &name, std::string *pool_name,
+                        std::string *namespace_name, std::string *image_name) {
+  // parse names like:
+  // '', 'rbd/', 'rbd/ns/', 'rbd//image', 'rbd/image', 'rbd/ns/image'
```
Nit: `rbd//image` wouldn't be valid under the new scheme, correct?

trociny replied:
Yes, it will always come from the scheduler module as `rbd/image`, though the scheduler module "understands" both ways. It does not hurt but ok I will remove this redundancy here.

dillaman commented:
I'm sure the dashboard will eventually want to integrate this support, which means we also will eventually need to add Python bindings for adding/removing/listing schedules. We can move the code when the time comes.

jan--f commented:
> May be @jan--f will have some comments about this? Do you have a command to list all schedules recursively? How does `ls` output looks in your case?

Apologies for the delay. 
My initial motivation for having a recursive and non-recusive variant of a `get/ls` command was that in the CephFS case I intended for them to display different information. Naming-wise I don't feel too strongly about either way. `ls` and `ls -R` certainly feel familiar.
The information in the recursive variant is supposed to be quite terse and make it clear where schedules are defined and display some limited info in them.
The on-recursive variant would then print all (or most) of the interesting metadata. Things like first run, last run, maybe creation time and such. Everything that would be overload in a recursive listing. Maybe it makes sense to sync up on how this output should look for the various variants?

trociny commented:
@jan--f It seems like in non-recursive  variant you want to combine what we have in `ls` + `status`.

We have `status` command to display information about the next scheduled snapshot for an image. So right now the output for `status` is a table with two columns: "schedule time" and "image". It might be extended in future, e.g. to provide info about the last snapshot run for an image (right now we don't store this information, and it is not shown).

While `ls` command is intended to show only "configuration" information.

I am not sure about the idea to combine `ls` and `status` in our case. I like when they are separated. @dillaman what do you think?

jan--f commented:
hmm then I suppose we could go with ls/ls -R and status (though I still like get for this).
Though I wonder about the need for a simple ls.

1. For status or get its clear: a user wants to see all data about a snapshot schedules.
2. ls -R or whatever the recusive listing will be called: also clear, show me all schedules in this hierarchy.
3. naked ls: show me schedules for this item (dir or image or so) but with limited information.

I think my question would be, why would a user use option 3 over option 1?

trociny commented:
@jan--f it seems @dillaman had similar thoughts about not necessity of non-recursive `ls`.

Ok, I will try to explain my reason to have both, and if you still think it is not good enough I am fine with having only one (recursive) variant.

Let's consider a configuration: a schedule on global level + a schedule on "rbd/ns1" namespace level. Now if I run `ls` (non-recursive) for "rbd" pool level it will return ENOENT because nothing is configured on this level.  But when I run `ls -R` for "rbd" pool level, it will show that on this level the global setting is applied and will also show overrides on levels below (a schedule on "rbd/ns1" namespace level for this example).

I imagine non-recursive `ls` may be useful when you are configuring schedules and what to know what exactly settings are on this level. Sure, you can figure this out from the recursive output, but this is a bit of work. Additionally, you could use the non-recursive `ls` in scripts/tests to check the configuration for a particular level.

So non-recursive `ls` is to see what I have configured on a particular level, while `ls -R` is to see how images below a particular level are going to be scheduled.

For me it looks useful to have both, though I don't know how useful in the wild, and as I said, I am fine with having only the recursive variant.

dillaman commented:
> For me it looks useful to have both, though I don't know how useful in the wild, and as I said, I am fine with having only the recursive variant.

The recursive listing has grown on me once you put it in terms of "ls" vs "ls -R". 

Semi-related: were you planning to also drop the current MGR "list" command and rename "dump" to "list?

trociny commented:
> The recursive listing has grown on me once you put it in terms of "ls" vs "ls -R".

Just to make sure. You vote for just making "ls" do what currently "ls -R" does and dropping "-R" flag?

> Semi-related: were you planning to also drop the current MGR "list" command and rename "dump" to "list?

Yes. If we leave only "recursive listing" I am going to do this.

dillaman commented:
> > The recursive listing has grown on me once you put it in terms of "ls" vs "ls -R".
> 
> Just to make sure. You vote for just making "ls" do what currently "ls -R" does and dropping "-R" flag?

Sorry, I am saying I am fine w/ your current "ls" and "ls -R" split in the CLI (i.e. keep it as it is).

trociny commented:
> Sorry, I am saying I am fine w/ your current "ls" and "ls -R" split in the CLI (i.e. keep it as it is).

Ok. And you would like I drop the current MGR "list" command, rename "dump" to "list" and use it for both rbd "ls" and "ls -R"? (I am fine with this, just want to make sure).

jan--f commented:
Ok just to make sure I understand too :) We'll go with `ls` and `ls -R` for listing (recursive optional) and `status` or `get` for detailed info on a schedules for one item (image or directory for the fs case)?

trociny commented:
> Ok just to make sure I understand too :) We'll go with `ls` and `ls -R` for listing (recursive optional) and `status` or `get` for detailed info on a schedules for one item (image or directory for the fs case)?

In my current implementation `status` returns status (the next scheduled time) for all images under this level (global, pool, namespace, image). So, if you run `... status --pool rbd` it will output all scheduled images in rbd pool, if you run `... status --image rbd/test` it will output the status for this image only.

jan--f commented:
> In my current implementation `status` returns status (the next scheduled time) for all images under this level (global, pool, namespace, image). So, if you run `... status --pool rbd` it will output all scheduled images in rbd pool, if you run `... status --image rbd/test` it will output the status for this image only.


Well at some point things will diverge for rbd and cephfs so I think that is fine. But there is no other get-like command that returns comprehensive information on schedules for a particular item, right?

trociny commented:
> But there is no other get-like command that returns comprehensive information on schedules for a particular item, right?

There is no any other command. I suppose the `status` command may be extended in future to provide more information.

trociny commented:
updated:  `ceph rbd mirror snapshot schedule dump` renamed to `list` and `rbd` tweaked accordingly. Also added the new `rbd` commands to `rbd` man page.
